[WIP] Emit metric for seconds left until expiration for loaded certs

PeterL328 opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: Emit metric for seconds left until expiration for loaded certs
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

PeterL328 commented:
/retest

PeterL328 commented:
@ggreenway Do you mind taking a look at my PR? 😄

JuniorHsu commented on source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:
```diff
@@ -541,15 +541,25 @@ Envoy::Ssl::CertificateDetailsPtr DefaultCertValidator::getCaCertInformation() c
 
+absl::optional<std::pair<std::string, uint64_t>>
+DefaultCertValidator::getSecondsUntilExpirationOfCert() const {
+  auto tmp = Utility::getSecondsUntilExpiration(ca_cert_.get(), time_source_);
```
`tmp` needs better naming

JuniorHsu commented on source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:
```diff
@@ -41,9 +41,9 @@ namespace TransportSockets {
-    : config_(config), stats_(stats), time_source_(time_source) {
+    const std::string cert_name, const Envoy::Ssl::CertificateValidationContextConfig* config,
+    SslStats& stats, TimeSource& time_source)
+    : config_(config), stats_(stats), time_source_(time_source), cert_name_(cert_name) {
```
probably modify the data member ordering which matches the constructor's argumetns

JuniorHsu commented on source/extensions/transport_sockets/tls/context_config_impl.cc:
```diff
@@ -91,22 +96,28 @@ getCertificateValidationContextConfigProvider(
   }
   default:
-    return nullptr;
+    return {"", nullptr};
```
EMPTY_STRING, you may needs to include some headers

JuniorHsu commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -573,6 +579,25 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+                                       ca_cert_expiration.value().second);
+  }
+  // Add TLS certs' expiration time.
+  for (auto& ctx : tls_contexts_) {
```
const auto& if possible

jmarantz replied:
Please use explicit type. It doesn't look like it can be const though.

JuniorHsu commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -573,6 +579,25 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  }
+  // Add TLS certs' expiration time.
+  for (auto& ctx : tls_contexts_) {
+    auto tmp = Utility::getSecondsUntilExpiration(ctx.cert_chain_.get(), time_source_);
```
rename tmp

JuniorHsu commented on source/server/server.cc:
```diff
@@ -302,8 +302,30 @@ void InstanceImpl::updateServerStats() {
+      sslContextManager().getSecondsUntilExpirationOfCerts();
+  for (const auto& cert_name_seconds_until_expiration_pair :
+       cert_name_seconds_until_expiration_pairs) {
+    std::string cert_name = cert_name_seconds_until_expiration_pair.first;
```
const ref

zuercher commented:
/assign-from @envoyproxy/first-pass-reviewers

PeterL328 commented on envoy/ssl/context.h:
```diff
@@ -20,6 +20,12 @@ class Context {
+  /**
+   * @return a vector of pairs of cert name and seconds until expiration for all loaded certs.
+   */
+  virtual std::vector<std::pair<std::string, uint64_t>>
```
Replace std::pair usages with struct for clearness to avoid reading code with `.first`, `.second`

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
> CC @envoyproxy/dependency-shepherds: Your approval is needed for changes made to `(bazel/.*repos.*\.bzl)|(bazel/dependency_imports\.bzl)|(api/bazel/.*\.bzl)|(.*/requirements\.txt)|(.*\.patch)`. envoyproxy/dependency-shepherds assignee is @RyanTheOptimist
> 
> 🐱

sorry, I think I accidentally did a `git merge upstream/main` which pulled in several other changes. I reverted those.

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
Seems like getting an unrelated error from clang-tidy. I'm not sure how to fix it but was able to get green CI from retest though takes a few tries.

PeterL328 commented:
/retest

PeterL328 commented:
clang-tidy error in CI shows the following:
```
Error while processing /source/envoy/ssl/certificate_validation_context_config.h.

/build/bazel_root/base/execroot/envoy/./source/common/common/base_logger.h:7:10: error: 'spdlog/spdlog.h' file not found [clang-diagnostic-error]
#include "spdlog/spdlog.h"
.
.
.
```
I did add some code in `certificate_validation_context_config.h` but not sure how it relates with `base_logger.h` and not sure why `'spdlog/spdlog.h' file not found`

RyanTheOptimist commented:
/assign @ggreenway 
since this is TLS related

ggreenway commented on source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:
```diff
@@ -282,6 +282,24 @@ std::string SPIFFEValidator::extractTrustDomain(const std::string& san) {
 
+absl::optional<Ssl::CertificateNameExpiration>
+SPIFFEValidator::getSecondsUntilExpirationOfCert() const {
+  // TODO(yleng): With the current interface, we cannot pass the multiple cert information.
```
I think it would be more correct to return the value for the cert that expires earliest

PeterL328 replied:
Updated to return a vector based on below comment: https://github.com/envoyproxy/envoy/pull/27687#discussion_r1243928956

ggreenway commented on envoy/ssl/context.h:
```diff
@@ -13,13 +13,23 @@ namespace Ssl {
 
+struct CertificateNameExpiration {
+  std::string name;
+  uint64_t expiration;
```
This interface would be nicer if it held a `std::chrono::duration`

PeterL328 replied:
Sounds good. Updated

ggreenway commented on envoy/ssl/context.h:
```diff
@@ -13,13 +13,23 @@ namespace Ssl {
+  /**
+   * @return a vector of struct with certificate names and expiration times for all loaded certs.
+   */
+  virtual std::vector<CertificateNameExpiration> getSecondsUntilExpirationOfCerts() const PURE;
```
```suggestion
  virtual std::vector<CertificateNameExpiration> getExpirationOfCerts() const PURE;
```

ggreenway commented on envoy/ssl/context_manager.h:
```diff
@@ -32,6 +32,11 @@ class ContextManager {
+  /**
+   * @return a vector of struct with certificate names and expiration times for all loaded certs.
+   */
+  virtual std::vector<CertificateNameExpiration> getSecondsUntilExpirationOfCerts() const PURE;
```
```suggestion
  virtual std::vector<CertificateNameExpiration> getExpirationOfCerts() const PURE;
```

ggreenway commented on source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:
```diff
@@ -105,6 +105,11 @@ class CertValidator {
+   * Get a struct with CA cert name and the seconds until the CA cert expires.
+   */
+  virtual absl::optional<Ssl::CertificateNameExpiration>
+  getSecondsUntilExpirationOfCert() const PURE;
```
```suggestion
  getExpirationOfCert() const PURE;
```

ggreenway commented on source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:
```diff
@@ -105,6 +105,11 @@ class CertValidator {
+  /**
+   * Get a struct with CA cert name and the seconds until the CA cert expires.
+   */
+  virtual absl::optional<Ssl::CertificateNameExpiration>
```
This should return a vector; some validators have multiple CA certs

ggreenway commented on source/server/server.cc:
```diff
@@ -302,8 +302,30 @@ void InstanceImpl::updateServerStats() {
+       cert_name_seconds_until_expiration_structs) {
+    const std::string& cert_name = cert_name_seconds_until_expiration_struct.name;
+    uint64_t expiration_date = cert_name_seconds_until_expiration_struct.expiration;
+    if (cert_stats_map_.find(cert_name) == cert_stats_map_.end()) {
```
Stats for certs that no longer exist (due to reconfiguration) need to be deleted.

I wonder if this code should move into the tls transport socket instead of being here (even though the existing daysUntil code is here). The daysUntil needed to be in a common place because it had to look at all the certs in the system and find the one with the shortest expiration.

This new stat is per-cert, and so it feels like it should be attached to the config object that holds it. That would automatically handle deletion when config goes away, and the names could be generated under the existing stat prefix for the transport socket. @jmarantz wdyt?

jmarantz replied:
@ggreenway that sounds right to me.

ggreenway replied:
@PeterL328 please move all the code into the tls transport socket as described.

PeterL328 replied:
Will make the changes as described above and let you know when ready to review. Thanks for feedback.

PeterL328 replied:
Moved code into tls transport socket as described.

ggreenway reviewed (changes requested):
/wait

jmarantz commented:
RE tidy not finding spdlog.h: I believe I've seen this before. The cause is BUILD files that don't fully list required dependencies. Normal compiles work because of transitive dependencies. But, for reasons I don't understand, clang-tidy is stricter. So a potential remedy is adding direct BUILD dependencies as indicated by the detailed tidy logs.

PeterL328 commented:
fyi, going on a short vacation and be back to work on this end of this month. @ggreenway

ggreenway commented:
> fyi, going on a short vacation and be back to work on this end of this month. @ggreenway

Sounds good; ping me when you get back and this is ready for review. And if stale-bot closes it I'll re-open.

/wait

PeterL328 commented:
Addressed feedback.
Ready for another round of review :grin:

PeterL328 commented:
/retest

PeterL328 commented:
/retest

ggreenway commented on changelogs/current.yaml:
```diff
@@ -16,5 +16,9 @@ new_features:
     added ECDS support for :ref:` downstream network filters<envoy_v3_api_field_config.listener.v3.Filter.config_discovery>`.
+- area: tls
+  change: |
+    Added new metric for emitting seconds until expirations of TLS & CA certificates whcih are rooted at the cert.<CERT_NAME>. namespace.
```
```suggestion
    Added new metric for emitting seconds until expirations of TLS & CA certificates which are rooted at the cert.<CERT_NAME>. namespace.
```

Also, I don't think that's the right namespace anymore

PeterL328 replied:
also updated docs

ggreenway commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -595,6 +601,33 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+    const std::string& cert_name = cert.name;
+    uint64_t expiration_date = cert.expiration.count();
+    if (cert_stats_map_.find(cert_name) == cert_stats_map_.end()) {
+      cert_stats_map_[cert_name].scope = store.createScope(absl::StrCat("cert.", cert_name, "."));
```
You don't need a separate scope for each stat; they should all be in `scope_` which already exists

PeterL328 replied:
Removed this part of the code since we don't create the stat here anymore.

ggreenway commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -595,6 +601,33 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+      cert_validator_->getExpirationOfCerts();
+
+  // Add TLS certs' expiration time.
+  for (const auto& ctx : tls_contexts_) {
```
Instead of discovering all the certs here, can you create the stats for them when the cert is added (and delete the stat when the cert is deleted)?

With this approach, you'll need to do a diff and figure out which ones have been deleted and remove them from the map. And this will all have to be computed every time the timer fires.

PeterL328 replied:
Yea sounds good. I moved creation of certStat for TLS certs into the TlsContext struct and creation of certStat for CA certs into the cert validator which should delete the stat when the cert is deleted. 
Though I don't expect CA certs to be deleted(?)

ggreenway commented on source/server/server.cc:
```diff
@@ -304,6 +304,9 @@ void InstanceImpl::updateServerStats() {
 void InstanceImpl::flushStatsInternal() {
   updateServerStats();
+  // Update certificate stats for each context.
+  ssl_context_manager_->iterateContexts(static_cast<std::function<void(Envoy::Ssl::Context&)>>(
```
Move this to `updateServerStats()`; I think that will be a little bit cleaner

PeterL328 replied:
Good idea. Moved

ggreenway reviewed (changes requested):
/wait

jmarantz commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -595,6 +601,33 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  for (const auto& cert : certs_expiration_time) {
+    const std::string& cert_name = cert.name;
+    uint64_t expiration_date = cert.expiration.count();
+    if (cert_stats_map_.find(cert_name) == cert_stats_map_.end()) {
```
drive-by perf nit: use map::insert and capture the return pair, to save a hash lookup.
```
auto insertion = cert_stats_map_.insert(cert_name);
if (insertion.second) {
  insertion.first.second = new value
}
```
this would replace the `find` and the `operator[]` both of which will do the same hash lookup.

PeterL328 replied:
Thanks. I used above approach in `source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc`.

PeterL328 commented:
Commenting to bring out of stale. Will work on it next week. Had some other high priority work last few weeks so wasn't able to work on it.

PeterL328 commented:
Sorry for delay, @ggreenway  I addressed your feedback. Let me know what you think. :)

ggreenway commented on source/common/ssl/certificate_validation_context_config_impl.cc:
```diff
@@ -19,10 +19,11 @@ static const std::string INLINE_STRING = "<inline>";
 CertificateValidationContextConfigImpl::CertificateValidationContextConfigImpl(
     const envoy::extensions::transport_sockets::tls::v3::CertificateValidationContext& config,
-    Api::Api& api)
+    Api::Api& api, const std::string& name)
```
```suggestion
    Api::Api& api, const std::string& ca_cert_name)
```

ggreenway commented on source/common/ssl/certificate_validation_context_config_impl.cc:
```diff
@@ -47,9 +48,9 @@ CertificateValidationContextConfigImpl::CertificateValidationContextConfigImpl(
 CertificateValidationContextConfigImpl::create(
     const envoy::extensions::transport_sockets::tls::v3::CertificateValidationContext& context,
-    Api::Api& api) {
+    Api::Api& api, const std::string& name) {
```
ditto

ggreenway commented on source/common/ssl/tls_certificate_config_impl.cc:
```diff
@@ -28,11 +28,13 @@ static const std::string INLINE_STRING = "<inline>";
     const envoy::extensions::transport_sockets::tls::v3::TlsCertificate& config,
-    Server::Configuration::TransportSocketFactoryContext& factory_context, Api::Api& api)
+    Server::Configuration::TransportSocketFactoryContext& factory_context, Api::Api& api,
+    const std::string& name)
```
```suggestion
    const std::string& certificate_name)
```

ggreenway commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -592,6 +599,25 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+  std::vector<Ssl::CertificateNameExpiration> ca_certs_expiration_time =
+      cert_validator_->getExpirationOfCerts();
+  for (const auto& ca_cert : ca_certs_expiration_time) {
+    cert_validator_->setExpirationOnCertStatsGivenCertName(ca_cert.name, ca_cert.expiration);
```
I don't understand what's going on here. It looks like you're asking the validator for something, then giving that right back to the validator. Why not have a single call on the validator to have it do whatever needs to be done?

PeterL328 replied:
Yea I agree. Maybe can just have a single function inside the validator. Maybe refreshCertStats() ?

PeterL328 replied:
Replaced `getExpirationOfCerts ()` and `setExpirationOnCertStatsGivenCertName()` with `refreshCertStatsWithExpirationTime()`

ggreenway commented on source/extensions/transport_sockets/tls/stats.cc:
```diff
@@ -14,6 +14,11 @@ SslStats generateSslStats(Stats::Scope& store) {
 }
 
+CertStats generateCertStats(Stats::Scope& scope, std::string cert_name) {
+  std::string prefix_cert_name("ssl." + cert_name + ".");
```
```suggestion
  std::string prefix_cert_name(absl::StrCat("ssl.", cert_name, "."));
```

PeterL328 replied:
I updated naming to `ssl.certificate.<cert_name>.`

ggreenway commented on source/extensions/transport_sockets/tls/stats.cc:
```diff
@@ -14,6 +14,11 @@ SslStats generateSslStats(Stats::Scope& store) {
 
+CertStats generateCertStats(Stats::Scope& scope, std::string cert_name) {
+  std::string prefix_cert_name("ssl." + cert_name + ".");
+  return {ALL_CERT_STATS(POOL_GAUGE_PREFIX(scope, prefix_cert_name))};
```
Please add tag-extraction rules for these certs here: https://github.com/envoyproxy/envoy/blob/cfdc99a00d43c6c08c22229c7053410446e51a3a/source/common/config/well_known_names.cc#L201

And add some tests for them

PeterL328 replied:
Got it. Added

ggreenway commented on source/server/admin/server_info_handler.cc:
```diff
@@ -18,17 +18,18 @@ Http::Code ServerInfoHandler::handlerCerts(Http::ResponseHeaderMap& response_hea
-      *cert_chain = *cert_details;
-    }
-  });
+  server_.sslContextManager().iterateContexts(static_cast<std::function<void(const Ssl::Context&)>>(
```
Why the static_cast of a std::function?

PeterL328 replied:
See explanation here
https://github.com/envoyproxy/envoy/pull/27687/files#r1332712545

ggreenway commented on source/server/server.cc:
```diff
@@ -301,6 +301,10 @@ void InstanceImpl::updateServerStats() {
       stats_store_.symbolTable().getRecentLookups([](absl::string_view, uint64_t) {}));
+
+  // Update certificate stats for each context.
+  ssl_context_manager_->iterateContexts(static_cast<std::function<void(Envoy::Ssl::Context&)>>(
```
ditto

PeterL328 replied:
The reason I have a static_cast of a std::function is due to ambiguity of iterateContexts since I have overloaded the `iterateContexts` that also takes in a `std::function<void(Context&)>`. which has non-const Context (see context_manager.h). 

The reason I needed to overloaded it is because I use it here and `context.updateCertStats()` is not const.

jmarantz replied:
Can I suggest you instead rename the overload? Forcing this usage is a bit awkward.

PeterL328 replied:
I removed the polling here so shouldn't be a concern.

ggreenway reviewed (changes requested):
One more high-level question: instead of seconds-until-expiration, what if we published it as the time in seconds-since-unix-epoch of the expiration. This has two advantages: it's not constantly changing and taking up more space in a TSDB, and envoy doesn't have to keep updating the value (it gets set once when we load the cert and read it's expiration time). I know that makes it different than the existing envoy metric for time until next cert expiration, but seems worth considering.

/wait

PeterL328 commented:
> One more high-level question: instead of seconds-until-expiration, what if we published it as the time in seconds-since-unix-epoch of the expiration. This has two advantages: it's not constantly changing and taking up more space in a TSDB, and envoy doesn't have to keep updating the value (it gets set once when we load the cert and read it's expiration time). I know that makes it different than the existing envoy metric for time until next cert expiration, but seems worth considering.
> 
> /wait
@ggreenway 
Seems if we switch to `seconds-since-unix-epoch of the expiration` it would be similar to viewing the certs from the /certs endpoint? 
I think our use case is we want to have alerts setup that points to specific certs to notify if the certs are expiring soon. 
Do you think we can move forward with current approach and we can iterate on it in next PRs if we want to tweak it?

phlax commented:
/retest

ggreenway commented:
> Seems if we switch to `seconds-since-unix-epoch of the expiration` it would be similar to viewing the certs from the /certs endpoint?
> I think our use case is we want to have alerts setup that points to specific certs to notify if the certs are expiring soon.

Can you setup an alert of the form `if expiration_time_stat - current_time < threshold`? Many alerting systems support things like this.

> Do you think we can move forward with current approach and we can iterate on it in next PRs if we want to tweak it?

Once we start publishing a metric, changing it to something completely different is more difficult, so I think we should figure out what we want it to look like now.

PeterL328 commented:
Got some time to work on this again.
Yeah sounds good. I'll update to emit just the seconds-since-unix-epoch of the expiration.

PeterL328 commented:
Bringing out of stale.
Will work on this next week.

RyanTheOptimist commented:
/wait
Needs a main merge

PeterL328 commented:
@ggreenway 
Sorry didn't get change to work on this until recently.
Updated PR to address previous concerns. 
Metrics is now giving seconds since UNIX epoch of expiration of certs

PeterL328 commented:
/retest

adisuissa commented:
Seems that it needs a main merge.
/wait

PeterL328 commented:
Rebased with main

adisuissa commented:
@RyanTheOptimist @ggreenway PTAL.

RyanTheOptimist commented:
I was originally added as a deps reviewer and it appears this no longer includes deps changes, so I will remove that tag and un-assign myself.

That being said, as per the advice here https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md, please do not rebase PRs. Instead, please use `git merge`. That makes it much easier to review.

Also, it appears to still need a main merge.

/wait

PeterL328 commented:
merged with main

phlax commented on changelogs/current.yaml:
```diff
@@ -364,6 +364,11 @@ new_features:
     Added support for ``%CONNECTION_ID%`` command operator for UDP session access log.
+- area: tls
+  change: |
+    Added new metric for emitting seconds since UNIX epoch of expirations of TLS & CA certificates.
```
```suggestion
    Added new metric for emitting seconds since UNIX epoch of expirations of TLS and CA certificates.
```

phlax commented on changelogs/current.yaml:
```diff
@@ -364,6 +364,11 @@ new_features:
+- area: tls
+  change: |
+    Added new metric for emitting seconds since UNIX epoch of expirations of TLS & CA certificates.
+    They are rooted at cluster.<cluster_name>.ssl.certificate.<cert_name>.
+    and at listener.<address>.ssl.certificate.<cert_name>. namespace.
```
```suggestion
    They are rooted at ``cluster.<cluster_name>.ssl.certificate.<cert_name>.``
    and at ``listener.<address>.ssl.certificate.<cert_name>.`` namespace.
```

phlax replied:
looks like this isnt resolved in current version - also this seems to be deleting unrelated change logs

PeterL328 replied:
thanks updated. Must have accidentally removed it on resolving main merge conflict.

phlax commented on docs/root/_include/cert_stats.rst:
```diff
@@ -0,0 +1,5 @@
+   :header: Name, Type, Description
+   :widths: 1, 1, 2
+
+   seconds_since_epoch_of_expiration, Gauge, Number of seconds since UNIX epoch of the expiration date of the certificate.
```
```suggestion
   ``seconds_since_epoch_of_expiration``, Gauge, Number of seconds since UNIX epoch of the expiration date of the certificate.
```

phlax commented on docs/root/configuration/listeners/stats.rst:
```diff
@@ -41,6 +41,15 @@ The following TLS statistics are rooted at *listener.<address>.ssl.*:
 
+.. _config_listener_stats_certs:
+
+TLS & CA certificates
```
```suggestion
TLS and CA certificates
```

phlax commented on docs/root/configuration/listeners/stats.rst:
```diff
@@ -41,6 +41,15 @@ The following TLS statistics are rooted at *listener.<address>.ssl.*:
+TLS & CA certificates
+---------------------
+
+TLS & CA certificate statistics are rooted in the *listener.<address>.ssl.certificate.<cert_name>.*:
```
```suggestion
TLS and CA certificate statistics are rooted in the ``listener.<address>.ssl.certificate.<cert_name>.``:
```

phlax commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -266,6 +266,15 @@ If TLS is used by the cluster the following statistics are rooted at *cluster.<n
 
+.. _config_cluster_manager_cluster_stats_certs:
+
+TLS & CA certificates
```
```suggestion
TLS and CA certificates
```

phlax commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -266,6 +266,15 @@ If TLS is used by the cluster the following statistics are rooted at *cluster.<n
+TLS & CA certificates
+---------------------
+
+TLS & CA certificate statistics are rooted in the *cluster.<name>.ssl.certificate.<cert_name>.*:
```
```suggestion
TLS and CA certificate statistics are rooted in the ``cluster.<name>.ssl.certificate.<cert_name>.``:
```

phlax reviewed (commented):
thanks for working on this @PeterL328 - docs review below

PeterL328 commented:
@phlax  Thanks
Addressed docs feedback

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

ggreenway commented on docs/root/_include/cert_stats.rst:
```diff
@@ -0,0 +1,5 @@
+   :header: Name, Type, Description
+   :widths: 1, 1, 2
+
+   ``seconds_since_epoch_of_expiration``, Gauge, Number of seconds since UNIX epoch of the expiration date of the certificate.
```
I think the name of the metric could be shorted to `expiration_unix_time`. I think "UNIX time" is widely-enough understood that this will be a more precise and concise name.

PeterL328 replied:
Sounds good. Let's go with `expiration_unix_time`

ggreenway commented on envoy/secret/secret_provider.h:
```diff
@@ -78,5 +78,15 @@ using GenericSecretConfigProvider =
     SecretProvider<envoy::extensions::transport_sockets::tls::v3::GenericSecret>;
 using GenericSecretConfigProviderSharedPtr = std::shared_ptr<GenericSecretConfigProvider>;
 
+struct TlsCertificateConfigProviderWithName {
```
I don't think these types need to be defined in this interface, as they aren't used here. I think they could be defined in context_impl.h, probably near where the functions that return them are declared.

ggreenway commented on envoy/ssl/context.h:
```diff
@@ -13,13 +15,23 @@ namespace Ssl {
 
+struct CertificateNameExpiration {
+  std::string name;
+  std::chrono::duration<uint64_t> expiration;
```
```suggestion
  std::chrono::seconds expiration;
```

ggreenway commented on source/server/server.cc:
```diff
@@ -276,6 +276,10 @@ void InstanceBase::updateServerStats() {
+
+  // Update certificate stats for each context.
+  ssl_context_manager_->iterateContexts(static_cast<std::function<void(Envoy::Ssl::Context&)>>(
+      [&](Envoy::Ssl::Context& context) -> void { context.updateCertStats(); }));
```
Is it necessary to poll for updates here? I think you can set the correct stats whenever the secret-update callback is run in the tls_context.

PeterL328 replied:
Good point. We don't need to poll for updates. 
Updated.

jmarantz commented on source/common/config/well_known_names.cc:
```diff
@@ -207,6 +207,11 @@ TagNameValues::TagNameValues() {
+  // Signing algorithms and ciphers have the same pattern so they use the same regex.
+  // listener.[<address>.]ssl.sigalgs.(<algorithm>)
+  addRe2(TLS_CERTIFICATE, R"(^<LISTENER_OR_CLUSTER_WITH_NAME>\.ssl\.certificate(\.(<TAG_VALUE>))$)",
+         ".ssl.certificate.");
```
it looks like we can use a tokenized stat-matcher for this, which will have les overhead. Have you tried it?

PeterL328 replied:
Haven't tried it yet. Will replace with tokenized one also update the comment above it which is copied wrong.

PeterL328 replied:
Sorry I tried tokenizer but it doesn't work. I think the issue is that the tokenizer doesn't transform the metric.
i.e we'll need 
"listener.0.0.0.0_0.ssl.certificate.server_cert.expiration_unix_time" -> "listener.ssl.certificate".

PeterL328 replied:
Let me know if we are fine with moving forward like this for now.

jmarantz commented on source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:
```diff
@@ -551,15 +551,26 @@ Envoy::Ssl::CertificateDetailsPtr DefaultCertValidator::getCaCertInformation() c
 }
 
+void DefaultCertValidator::refreshCertStatsWithExpirationTime() {
+  auto expiration_unix_time = Utility::getExpirationUnixTime(ca_cert_.get());
```
nit: use explicit type here, as the type is not visible from context.

PeterL328 replied:
Got it.

jmarantz commented on source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:
```diff
@@ -551,15 +551,26 @@ Envoy::Ssl::CertificateDetailsPtr DefaultCertValidator::getCaCertInformation() c
 
+void DefaultCertValidator::refreshCertStatsWithExpirationTime() {
+  auto expiration_unix_time = Utility::getExpirationUnixTime(ca_cert_.get());
+  if (expiration_unix_time.has_value()) {
```
if the expiration time has no value, does that mean it's a non-expiring cert? Should we emit something for that, like (say) '0'?

PeterL328 replied:
for non-expiring certs, we can set to something like max uint64_t? 
Good point.

jmarantz reviewed (commented):
lots of changes for this!

Why do we want this as a stat? This seems to be something that might be more efficiently emitted as a log. Is there a need to look at this in a time-series database historically?

jmarantz commented:
/wait for now as I want to confirm we really want this data in stats before I review the rest of this.

ggreenway commented:
I think it makes sense to put into stats because:

* Existing stat for next-to-expire cert is in stats
* It's much more common to setup alerting based on metrics than on logs
* Figuring out the current/latest state is more straightforward with metrics than with logs

jmarantz commented on source/extensions/transport_sockets/tls/utility.cc:
```diff
@@ -232,6 +232,22 @@ std::string Utility::getSubjectFromCertificate(X509& cert) {
+  }
+
+  // Obtain the expiration time as system time
+  auto expiration_time = Utility::getExpirationTime(*cert);
```
explicit type

jmarantz reviewed (commented):
flushing comments; will continue but you can respond to my comments in parallel.

PeterL328 commented:
ty for feedback. Will address soon

PeterL328 commented:
commenting to bring out of stale.

PeterL328 commented on source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:
```diff
@@ -282,6 +282,26 @@ std::string SPIFFEValidator::extractTrustDomain(const std::string& san) {
+    const absl::optional<uint32_t> expiration_unix_time =
+        Utility::getExpirationUnixTime(cert.get());
+    if (expiration_unix_time.has_value()) {
+      std::string cert_name = absl::StrCat(cert_name_, idx);
```
Good catch there. Yes, let's add an underscore between the cert name and idx.

PeterL328 commented:
Had some issues when running `git rebase origin/main` which pulled in changes not done by me and added to the diff (perhaps I didn't run it correctly)
I had to do a rebase.

PeterL328 commented:
/retest

PeterL328 commented:
Hi @ggreenway and @jmarantz,
I've addressed your feedback. Can you take a look? Sorry been a while since I worked on this. 
Hopefully we can work on this together to land this soon. 

Thanks!

ggreenway commented:
> I could be not remembering correctly, but I'm pretty sure we have had this for a really long time? I remember adding something like this many years ago. @ggreenway ?
> 
> /wait-any

Right now we have a metric for time until the soonest-expiring cert expires. This PR is adding per-cert metric.

ggreenway commented on docs/root/_include/cert_stats.rst:
```diff
@@ -0,0 +1,5 @@
+   :header: Name, Type, Description
+   :widths: 1, 1, 2
+
+   ``expiration_unix_time``, Gauge, Number of seconds since UNIX epoch of the expiration date of the certificate.
```
Suffix with `_seconds` to this name for clarity?

PeterL328 replied:
sg, 
renamed to `expiration_unix_time_in_seconds`

ggreenway commented on source/common/tls/cert_validator/default_validator.h:
```diff
@@ -111,10 +112,12 @@ class DefaultCertValidator : public CertValidator, Logger::Loggable<Logger::Id::
   SslStats& stats_;
   Server::Configuration::CommonFactoryContext& context_;
-
+  Stats::Scope& scope_;
```
I think you can create the stat in the constructor, and then not need to keep a scope reference

PeterL328 replied:
Good point. Removed

PeterL328 replied:
side note: for spiffe validator we'll keep the initializeCertExpirationStats call in the constructor since that's where the logic to populate the cert is located.

ggreenway replied:
Pass the scope reference as a parameter to initializeSslContexts` and then you don't need to hold the reference ever.

PeterL328 replied:
Done, removed storing references to scope_

ggreenway commented on source/common/tls/context_config_impl.cc:
```diff
@@ -25,33 +25,36 @@ namespace Tls {
-  std::vector<Secret::TlsCertificateConfigProviderSharedPtr> providers;
+  std::vector<TlsCertificateConfigProviderWithName> providers;
   if (!config.tls_certificates().empty()) {
+    int unnamed_cert_idx = 0;
```
```suggestion
    uint32_t unnamed_cert_idx = 0;
```

ggreenway commented on source/common/tls/context_config_impl.cc:
```diff
@@ -25,33 +25,36 @@ namespace Tls {
       }
-      providers.push_back(
+      providers.push_back(TlsCertificateConfigProviderWithName{
+          absl::StrCat("unnamed_cert_", unnamed_cert_idx++),
```
I'm worried that these names may conflict with another tls context that also has unnamed certs. Could we use something like the serial number or a hash of the cert, to prevent collisions?

ggreenway commented on source/common/tls/context_config_impl.cc:
```diff
@@ -106,24 +109,31 @@ getCertificateValidationContextConfigProvider(
-        .secretManager()
-        .createInlineCertificateValidationContextProvider(config.validation_context());
+    return CertificateValidationContextConfigProviderWithName{
+        "unnamed_ca_cert",
```
Ditto about collisions

ggreenway commented on source/common/tls/context_impl.cc:
```diff
@@ -572,6 +584,24 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
+    absl::optional<uint64_t> expiration_unix_time =
+        Utility::getExpirationUnixTime(ctx.cert_chain_.get());
+    if (expiration_unix_time.has_value()) {
+      absl::Status status = ctx.setExpirationOnCertStats(
```
Can this fail? If not, make the function return void

PeterL328 replied:
Sure let's update to void

ggreenway commented on source/common/tls/utility.cc:
```diff
@@ -419,6 +419,22 @@ Envoy::Ssl::ParsedX509NamePtr Utility::parseSubjectFromCertificate(X509& cert) {
 }
 
+absl::optional<uint64_t> Utility::getExpirationUnixTime(const X509* cert) {
+  if (cert == nullptr) {
```
You can simplify some code by never passing nullptr to this, removing this check, and make the result not be optional.

PeterL328 replied:
Was following the pattern in this file of handling nullptr inside the function and return `absl::optional<...>`. 
Perhaps we can keep it like `absl::optional<...>` for consistency with other methods like `getDaysUntilExpiration`, ? let me know if you disagree.

ggreenway reviewed (changes requested):
I think the structure of this is roughly correct now, but some details to sort out

/wait

PeterL328 commented on source/common/tls/stats.h:
```diff
@@ -32,6 +32,16 @@ struct SslStats {
+
+CertStats generateCertStats(Stats::Scope& scope, std::string cert_name);
+
+using CertStatsPtr = std::unique_ptr<CertStats>;
```
I'm using pointers to the CertStats struct because it contains reference members (Stats::Gauge&) making it non-copyable and non-assignable, which prevents us from reinitializing stats on already constructed TLS contexts when certificates change via SDS updates. i.e `ctx.createCertStats(...)`
Going to keep the pointer for now if you are fine with ^.

PeterL328 replied:
Sounds good. I've updated to store Gauge* and remove the struct

PeterL328 commented on envoy/ssl/context.h:
```diff
@@ -21,6 +23,11 @@ class Context {
+  /**
+   * @brief Updates the certificate stats for this context.
+   */
+  virtual void updateCertStats() PURE;
```
Removed

PeterL328 commented on source/common/tls/context_impl.cc:
```diff
@@ -572,6 +584,24 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
 
+void ContextImpl::updateCertStats() {
+  // Update CA cert expiration time.
+  cert_validator_->refreshCertStatsWithExpirationTime();
```
Yea, it's not needed here. I've repositioned this call also renamed this the method from `refreshCertStatsWithExpirationTime` to `initializeCertExpirationStats`

PeterL328 commented:
/retest

lizan commented:
/unreview

PeterL328 commented:
/retest

ramaraochavali commented:
@PeterL328 Is there a way for user to configure whether these stats should be generated? Where is that configuration? Or is this enabled by default?

ggreenway commented on source/common/tls/context_config_impl.h:
```diff
@@ -20,6 +20,16 @@ namespace Tls {
+};
+
+struct CertificateValidationContextConfigProviderWithName {
+  const std::string certificate_name;
+  Secret::CertificateValidationContextConfigProviderSharedPtr provider;
```
```suggestion
  const std::string certificate_name_;
  Secret::CertificateValidationContextConfigProviderSharedPtr provider_;
```

ggreenway commented on source/common/tls/context_impl.cc:
```diff
@@ -807,5 +828,14 @@ absl::Status TlsContext::checkPrivateKey(const bssl::UniquePtr<EVP_PKEY>& pkey,
+      Extensions::TransportSockets::Tls::generateCertStats(scope, cert_name));
+}
+
+void TlsContext::setExpirationOnCertStats(std::chrono::duration<uint64_t> duration) {
```
```suggestion
void TlsContext::setExpirationOnCertStats(std::chrono::seconds duration) {
```

ggreenway reviewed (changes requested):
/wait

PeterL328 commented on source/common/tls/context_config_impl.cc:
```diff
@@ -25,20 +26,46 @@ namespace Tls {
+      }
+
+      // Fall back to unnamed_cert_idx if we couldn't generate a hash-based name
+      if (cert_id.empty()) {
```
ah good catch there.

PeterL328 commented on envoy/ssl/context.h:
```diff
@@ -1,11 +1,13 @@
 #pragma once
 
+#include <chrono>
```
Reverted

PeterL328 commented on source/common/tls/cert_validator/default_validator.h:
```diff
@@ -111,10 +111,13 @@ class DefaultCertValidator : public CertValidator, Logger::Loggable<Logger::Id::
                                  std::string* error_details, uint8_t* out_alert);
 
+  void initializeCertExpirationStats(Stats::Scope& scope);
+  const std::string cert_name_;
```
Removed storing cert_name_ here and also in spiffe_validator

PeterL328 replied:
Don't believe it's actually used here anyways.

PeterL328 commented on source/common/tls/context_config_impl.cc:
```diff
@@ -25,33 +26,62 @@ namespace Tls {
 
-std::vector<Secret::TlsCertificateConfigProviderSharedPtr> getTlsCertificateConfigProviders(
+std::string generateCertificateHash(const std::string& cert_data) {
+  if (cert_data.empty()) {
```
Yea good point. I think at this point we can remove that check

PeterL328 commented on source/common/tls/context_impl.cc:
```diff
@@ -379,6 +385,9 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
   }
+
+  // Update certificate stats after all certificates are loaded
+  updateCertStats();
```
Removed updateCertStats and relocated it's logic near gauge creation

PeterL328 commented on source/common/tls/context_impl.h:
```diff
@@ -48,12 +48,14 @@ struct TlsContext {
   bssl::UniquePtr<SSL_CTX> ssl_ctx_;
   bssl::UniquePtr<X509> cert_chain_;
   std::string cert_chain_file_path_;
+  std::string cert_name_;
```
No I think we can remove it

PeterL328 commented:
@ggreenway Ty. Addressed feedback

PeterL328 commented:
retest

PeterL328 commented:
/retest

PeterL328 commented:
/retest

ggreenway commented:
@ramaraochavali I hear what you're saying, but it seems like 1 extra metric per configured certificate shouldn't be that much addition to existing metrics. I may be making bad assumptions, but I'd assume most cases where there are a large number of certificates that there would also be other config for each one to setup routes: possibly separate clusters or HCM's (with different stat_prefix), which would both have far more metrics. But please correct me if my assumptions are wrong.

ggreenway commented on source/common/tls/context_impl.cc:
```diff
@@ -212,6 +212,14 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
       }
+
+      // Create and set the certificate expiration gauge.
+      ctx.createCertExpirationGauge(scope, tls_certificate.certificateName());
+      if (ctx.cert_chain_.get() != nullptr) {
+        ctx.setExpirationOnCertStats(
+            std::chrono::seconds(Utility::getExpirationUnixTime(ctx.cert_chain_.get())));
+      }
```
I don't think you need two extra functions for this; just inline the code. Also, I don't think you need to a member to keep the pointer to the gauge; it is never used again. And let's also move the "there isn't a cert" case into `getExpirationUnixTime` to be the same as `getDaysUntilExpiration` in that regard.

```suggestion
      Stats::Gauge& expire = Extensions::TransportSockets::Tls::createCertificateExpirationGauge(scope, cert_name);
      expire.set(std::chrono::seconds(Utility::getExpirationUnixTime(ctx.cert_chain_.get())).count());
```

ggreenway commented on source/common/tls/utility.cc:
```diff
@@ -419,6 +419,18 @@ Envoy::Ssl::ParsedX509NamePtr Utility::parseSubjectFromCertificate(X509& cert) {
   return parseX509NameFromCertificate(cert, CertName::Subject);
 }
 
+uint64_t Utility::getExpirationUnixTime(const X509* cert) {
+  // Obtain the expiration time as system time
+  SystemTime expiration_time = Utility::getExpirationTime(*cert);
+
+  // Convert the time to duration since epoch
+  std::chrono::seconds duration_since_epoch =
+      std::chrono::duration_cast<std::chrono::seconds>(expiration_time.time_since_epoch());
+
+  // Convert the duration to uint64_t and return
+  return static_cast<uint64_t>(duration_since_epoch.count());
+}
```
```suggestion
std::chrono::seconds Utility::getExpirationUnixTime(const X509* cert) {
  if (certain == nullptr) {
    return std::chrono::seconds::max();
  }
  // Obtain the expiration time as system time
  SystemTime expiration_time = Utility::getExpirationTime(*cert);

  // Convert the time to duration since epoch
  return std::chrono::duration_cast<std::chrono::seconds>(expiration_time.time_since_epoch());
}
```

ggreenway commented on source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:
```diff
@@ -439,6 +444,28 @@ std::string SPIFFEValidator::extractTrustDomain(const std::string& san) {
+    // Add underscore between cert name and index to avoid collisions
+    std::string cert_name_with_idx = absl::StrCat(cert_name, "_", idx);
+
+    Stats::Gauge*& expiration_gauge = expiration_gauges_map_[cert_name_with_idx];
```
This map is never populated, and isn't needed. You don't need to keep a pointer to the gauge; just call `createCertificateExpirationGauge`, set the value, then you're done.

ggreenway commented on source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.h:
```diff
@@ -117,6 +118,7 @@ class SPIFFEValidator : public CertValidator, Logger::Loggable<Logger::Id::secre
   std::shared_ptr<SpiffeData> spiffe_data_;
   std::vector<SanMatcherPtr> subject_alt_name_matchers_{};
   std::unique_ptr<Filesystem::Watcher> file_watcher_;
+  absl::flat_hash_map<std::string, Stats::Gauge*> expiration_gauges_map_;
```
delete

ggreenway reviewed (changes requested):
You'll need to add test cases for the different metrics that will be created: server certificates, client certificates, default validator CA, spiffe CA.

/wait

ggreenway commented:
> This will enabled for all listeners, clusters etc for both sidecars and gateways. There will be many listeners/clusters in larger meshes. I agree there would be other metrics generated for them and this is one of them. Also it is not that much useful for short lived certs for example (as they are rotated automatically by tooling). Mainly useful for certs with long expiry. In the past we discussed about [#12780 (comment)](https://github.com/envoyproxy/envoy/issues/12780#issuecomment-679262123) . It would be good if we have config control over for which certs we need this. But I am good as is - let us see what others find out based on usage. We can add config if it becomes real problem.

Sounds good. There's also `stats_matcher` in https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#envoy-v3-api-msg-config-metrics-v3-statsconfig to configure removal of any stat that isn't useful for an installation.

PeterL328 commented on envoy/ssl/BUILD:
```diff
@@ -26,6 +26,7 @@ envoy_cc_library(
     name = "context_interface",
     hdrs = ["context.h"],
     deps = [
+        "//envoy/stats:stats_interface",
```
Reverted

PeterL328 commented:
@ggreenway  Added cert tests and addressed feedback. Ready for review.

PeterL328 commented:
/retest

PeterL328 commented:
/retest

PeterL328 commented:
CI failure seems unrelated to change. 
Test coverage below threshold for source/extensions/common/aws, likely fixed with main branch merge.

PeterL328 commented:
/retest

ggreenway commented on source/common/config/well_known_names.cc:
```diff
@@ -238,6 +238,13 @@ TagNameValues::TagNameValues() {
+  // cluster.[<cluster_name>.]ssl.certificate.(<cert_name>).expiration_unix_time_seconds
+  addRe2(
+      TLS_CERTIFICATE,
+      R"(^<LISTENER_OR_CLUSTER_WITH_NAME>\.ssl\.certificate(\.(<TAG_VALUE>)\.expiration_unix_time_seconds)$)",
```
In case we ever add another stat under this hierarchy (maybe usage count of a certificate for server certs?):
```suggestion
      R"(^<LISTENER_OR_CLUSTER_WITH_NAME>\.ssl\.certificate(\.(<TAG_VALUE>)\..*)$)",
```

PeterL328 replied:
sg, updated

ggreenway commented on test/common/stats/tag_extractor_impl_test.cc:
```diff
@@ -516,6 +516,16 @@ TEST(TagExtractorTest, DefaultTagExtractors) {
+  certificate_name.value_ = "server_cert";
+
+  listener_address.value_ = "0.0.0.0_0";
+  regex_tester.testRegex(
```
Please also add a test case for a cluster (as the regex should do both listener and cluster)

PeterL328 replied:
Added

ggreenway reviewed (changes requested):
This looks great! Thanks for all the iteration on it!

Just a couple small things, then it's ready to merge.

/wait

PeterL328 commented:
/retest

PeterL328 commented:
Thanks for the help reviewing this PR @ggreenway. Great to see it merged!
